### ISSUE TITLE
[Android][AIR 33.1.1.758] TextField and parts or DisplayList elements disappearing with direct renderMode and significate performance degradation

### ISSUE THREAD

**Issue by itlancer** [airsdk/Adobe-Runtime-Support#1630]
## Problem Description
`TextField` and parts of `DisplayList` elements disappearing with `direct` `renderMode` with Android devices and AIR 33.1.1.758.
Also significate performance degradation observed with multiple (10-30) `TextField` objects usage with complex AIR applications.
Also random crashes can be observed with complex AIR applications.

It has been tested with latest AIR 33.1.1.758 with multiple Android devices with different OS versions.
Same problem in all cases.
There is no such issues with AIR 33.1.1.743 and earlier.
Also there is no such issues with other platforms (Windows/macOS/iOS).

May be related links:
https://github.com/airsdk/Adobe-Runtime-Support/issues/1624
https://github.com/airsdk/Adobe-Runtime-Support/issues/1619
https://github.com/airsdk/Adobe-Runtime-Support/issues/1299

## Steps to Reproduce
Launch code below with any Android device and roll over mouse (or click) to red square button. Just to emulate "render" under the `TextField`.

Application example with sources attached.
[textfield_disappearing_bug.zip](https://github.com/airsdk/Adobe-Runtime-Support/files/8024076/textfield_disappearing_bug.zip)

```
package {
	import flash.display.Sprite;
	import flash.text.TextField;
	import flash.display.SimpleButton;
	
	public class TextFieldDisappearingBug extends Sprite {
		
		public function TextFieldDisappearingBug() {
			var upState:Sprite = new Sprite();
			upState.graphics.beginFill(0xff0000);
			upState.graphics.drawRect(0, 0, 200, 200);
			upState.graphics.endFill();

			var overState:Sprite = new Sprite();
			overState.graphics.beginFill(0x00ff00);
			overState.graphics.drawRect(0, 0, 200, 200);
			overState.graphics.endFill();

			var button:SimpleButton = new SimpleButton(upState, overState, overState, overState);
			button.x = 200;
			addChild(button);

			var text:TextField = new TextField();
			text.text = "Text";
			text.x = 200;
			addChild(text);
		}
	}
}
```

**Actual Result:**
Text partially or fully disappear as shown below:
![actual](https://user-images.githubusercontent.com/10899066/152994646-c674bf25-6639-4a05-b2e5-98f0621814b7.png)
Sometimes other parts of application (`Graphics`, `DisplayList` elements) visually corrupted. Sometimes parts of `TextField` objects will be cutted. Sometimes `TextField` objects showing/hiding with visual delay. Sometimes looks like `TextField` objects using incorrect size, position or other `TextFormat` settings. Sometimes with using complex AIR apps crash happens.

**Expected Result:**
Render will be correct. No crashes or visual artifacts.
![expected](https://user-images.githubusercontent.com/10899066/152995469-f6ed4d67-5897-4176-88bf-32c16c2f4fdb.png)

## Known Workarounds
none
*use AIR 33.1.1.743 or earlier

**Comment by ajwfrost** [comment]
FYI a fix for this will be coming out in our next release, we've got an updated runtime (64-bit ARM version attached here in case it helps) which works in CPU/GPU/Direct modes, with a couple of outstanding problems regarding any characters that don't take a normal text colour (i.e. emojis)... due to a disparity between how these are rendered in Android vs in OpenGL (and internally in AIR), they're appearing with the blue/red channels being swapped. We're looking to see if we can update this, but you'll see it (a) in direct mode, and (b) when drawing to BitmapData.
[Runtime.zip](https://github.com/airsdk/Adobe-Runtime-Support/files/8053490/Runtime.zip)

**Comment by mwelk** [comment]
This issue still exists in AIR 33.1.1.795.

**Comment by ajwfrost** [comment]
Hi @mwelk - we can't reproduce the issues reported here, via the latest test package (textfield_disappearing_bug2.zip). What exactly is it that you're seeing that goes wrong? - do you have a quick test case for it?

thanks

**Comment by mwelk** [comment]
@ajwfrost I've added this simple test case and if you rotate the device, the screen turns white without showing the content anymore.
[FlashMobileTest.zip](https://github.com/airsdk/Adobe-Runtime-Support/files/8244883/FlashMobileTest.zip)

**Comment by ajwfrost** [comment]
Thanks @mwelk  ... this looks like something we need to sort out asap then!

thanks

**Comment by mwelk** [comment]
Hey @ajwfrost, thanks for the quick update!
I can confirm that the issue is gone with your patch.

Thanks a lot!

**Comment by ajwfrost** [comment]
Great, thanks for confirming!

**Comment by itlancer** [comment]
@ajwfrost, with your latest patch over AIR 33.1.1.795 all issues gone.
Waiting for new release to close this issue.

**Comment by ajwfrost** [comment]
Thanks for confirming @itlancer  - so we've got one or two more changes due to land shortly and will then be creating an updated release...

**Comment by mwelk** [comment]
> It looks like we've found the culprit (not properly handling a resize process - we get a request for 0x0 as part of this which puts it into a bad state...). Below is an updated runtime file for the ARMv8 platform in case it's helpful for you to check against, just extract this into the AIR SDK folder "runtimes\air\android\device\arm8". [Runtime.zip](https://github.com/airsdk/Adobe-Runtime-Support/files/8254396/Runtime.zip)
> 
> Waiting for one final fix (for and Android camera roll issue) and then we'll kick off an updated release...
> 
> thanks

Hey @ajwfrost, do you now when we can expect the updated release or do you also have an updated runtime for ARMv7 which we can use until the final release gets published?

**Comment by ajwfrost** [comment]
Hi @mwelk  - it's just going through QA currently, I hope to get it published tomorrow. There's one issue been found so far but I think we'll ignore that and resolve it in a future update, this one is already later than I would have liked given some of the issues that are in the .795 build!

thanks

**Comment by mwelk** [comment]
Thanks a lot for the info!

**Comment by itlancer** [comment]
Finally fixed with latest AIR 33.1.1.821.
Thank you!
